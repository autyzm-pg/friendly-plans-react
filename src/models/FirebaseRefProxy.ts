import firebase, { RNFirebase } from 'react-native-firebase';

export const getPlanSubItemRef = (
  studentId: string,
  planId: string,
  planItemId: string,
  planSubItemId: string,
): RNFirebase.firestore.DocumentReference => getPlanSubItemsRef(studentId, planId, planItemId).doc(planSubItemId);

export const getPlanSubItemsRef = (
  studentId: string,
  planId: string,
  planItemId: string,
): RNFirebase.firestore.CollectionReference => getPlanItemRef(studentId, planId, planItemId).collection('subItems');

export const getPlanItemRef = (
  studentId: string,
  planId: string,
  planItemId: string,
): RNFirebase.firestore.DocumentReference => getPlanItemsRef(studentId, planId).doc(planItemId);

export const getPlanItemsRef = (studentId: string, planId: string): RNFirebase.firestore.CollectionReference =>
  getPlanRef(studentId, planId).collection('planItems');

export const getPlanRef = (studentId: string, planId: string): RNFirebase.firestore.DocumentReference =>
  getPlansRef(studentId).doc(planId);

export const getPlansRef = (studentId: string): RNFirebase.firestore.CollectionReference =>
  getStudentRef(studentId).collection('plans');

export const getStudentRef = (studentId: string): RNFirebase.firestore.DocumentReference =>
  getStudentsRef().doc(studentId);

export const getStudentsRef = (userId: string = getAuthenticatedUserId()): RNFirebase.firestore.CollectionReference =>
  getUserRef(userId).collection('students');

export const getUserRef = (userId: string): RNFirebase.firestore.DocumentReference => getUsersRef().doc(userId);

const getUsersRef = (): RNFirebase.firestore.CollectionReference => {
  return firebase.firestore().collection('users');
};

export const getAuthenticatedUserId = (): string => firebase.auth().currentUser!.uid;
