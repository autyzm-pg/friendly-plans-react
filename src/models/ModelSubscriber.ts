import { RNFirebase } from 'react-native-firebase';
import { OperationalError } from '../infrastructure/Errors';
import { SubscribableModel } from './SubscribableModel';

export class ModelSubscriber<T> {
  private unsubscribeCollectionUpdatesCallback: any = null;
  private unsubscribeElementUpdatesCallback: any = null;

  subscribeCollectionUpdates = (parent: SubscribableModel, callback: (elements: T[]) => void): void => {
    if (this.unsubscribeCollectionUpdatesCallback !== null) {
      throw new OperationalError(
        'Trying to subscribe on collection updates when there is a subscription already. Unsubscribe first.',
      );
    }
    this.unsubscribeCollectionUpdatesCallback = parent
      .getChildCollectionRef()
      .onSnapshot(this.parseCollectionDataToObjects(parent, callback));
  };

  unsubscribeCollectionUpdates = () => {
    if (this.unsubscribeCollectionUpdatesCallback === null) {
      throw new OperationalError(
        'Trying to unsubscribe on collection updates when there is no subscription. Subscribe first.',
      );
    }
    this.unsubscribeCollectionUpdatesCallback();
    this.unsubscribeCollectionUpdatesCallback = null;
  };

  subscribeElementUpdates = (element: SubscribableModel, callback: (element: T) => void): void => {
    if (this.unsubscribeElementUpdatesCallback !== null) {
      throw new OperationalError(
        'Trying to subscribe on element updates when there is a subscription already. Unsubscribe first.',
      );
    }
    this.unsubscribeElementUpdatesCallback = element
      .getRef()
      .onSnapshot(this.parseSingleElementDataToObject(element, callback));
  };

  unsubscribeElementUpdates = () => {
    if (this.unsubscribeElementUpdatesCallback === null) {
      throw new OperationalError(
        'Trying to unsubscribe on element updates when there is no subscription. Subscribe first.',
      );
    }
    this.unsubscribeElementUpdatesCallback();
    this.unsubscribeElementUpdatesCallback = null;
  };

  private parseCollectionDataToObjects = (
    parent: SubscribableModel,
    callback: (elements: T[]) => void,
  ): ((querySnapshot: RNFirebase.firestore.QuerySnapshot) => void) => {
    return querySnapshot => {
      const ChildType = parent.getChildType();
      const elements: T[] = querySnapshot.docs.map(document => {
        return Object.assign(new ChildType(), {
          id: document.id,
          ...document.data(),
        });
      });
      callback(elements);
    };
  };

  private parseSingleElementDataToObject = (
    element: SubscribableModel,
    callback: (element: T) => void,
  ): ((document: RNFirebase.firestore.DocumentSnapshot) => void) => {
    return document => {
      if (document.exists) {
        const updatedElement = Object.assign(new element.constructor(), {
          id: document.id,
          ...document.data(),
        });
        callback(updatedElement);
      }
    };
  };
}
