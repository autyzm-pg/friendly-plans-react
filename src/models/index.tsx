export * from './DialogProps';
export * from './User';
export * from './Student';
export * from './Plan';
export * from './PlanItem';
export * from './PlanSubItem';
