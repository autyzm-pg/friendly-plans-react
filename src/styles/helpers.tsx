import { ViewStyle } from 'react-native';
import { palette } from './palette';

export const getElevation = (elevation: number): ViewStyle => ({
  elevation,
  shadowColor: palette.shadow,
  shadowOffset: {
    width: 0,
    height: Math.round(1 + elevation / 4),
  },
  shadowOpacity: 0.17 + elevation / 100,
  shadowRadius: 1 + elevation / 2,
});
