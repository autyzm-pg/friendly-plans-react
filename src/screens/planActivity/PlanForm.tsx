import { Formik, FormikProps } from 'formik';
import React from 'react';
import { NavigationInjectedProps, withNavigation } from 'react-navigation';

import { Plan } from 'models';
import PlanFormControls from './PlanFormControls';

export interface PlanFormData {
  planInput: string;
}

const FORM_INITIAL_VALUES: PlanFormData = {
  planInput: '',
};

export class PlanForm extends React.PureComponent<NavigationInjectedProps> {
  async createPlan(planInput: string) {
    const { id } = this.props.navigation.getParam('student');

    const plan = await Plan.createPlan(id, planInput);

    this.props.navigation.setParams({ plan });
  }

  onSubmit = ({ planInput }: PlanFormData) => this.createPlan(planInput);

  renderFormControls = (props: FormikProps<PlanFormData>) => <PlanFormControls {...props} />;

  render() {
    return <Formik initialValues={FORM_INITIAL_VALUES} onSubmit={this.onSubmit} render={this.renderFormControls} />;
  }
}

export default withNavigation(PlanForm);
