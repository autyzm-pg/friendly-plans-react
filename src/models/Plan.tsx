import { RNFirebase } from 'react-native-firebase';

import { i18n } from 'locale';
import { PlanItemType } from './PlanItem';
import { getStudentsRef } from './Student';

export class Plan {
  name!: string;
  id!: string;
  studentId!: string;

  delete = (): Promise<void> => deletePlan(this);

  update = (changes: object): Promise<void> => updatePlan(this, changes);

  createPlanItem = (
    planItemType: PlanItemType,
  ): Promise<RNFirebase.firestore.DocumentReference> =>
    createPlanItem(this, planItemType);

  getPlanItemsRef = (): RNFirebase.firestore.CollectionReference =>
    getPlanItemsRef(this); 
}

// Private API below
const getPlanRef = (plan: Plan): RNFirebase.firestore.DocumentReference =>
  getStudentsRef()
    .doc(plan.studentId)
    .collection('plans')
    .doc(plan.id);

const deletePlan = (plan: Plan): Promise<void> => getPlanRef(plan).delete();

const updatePlan = (plan: Plan, changes: object): Promise<void> =>
  getPlanRef(plan).update(changes);

const getPlanItemsRef = (
  plan: Plan,
): RNFirebase.firestore.CollectionReference =>
  getPlanRef(plan).collection('planItems');

const createPlanItem = (
  plan: Plan,
  type: PlanItemType,
): Promise<RNFirebase.firestore.DocumentReference> =>
  getPlanItemsRef(plan).add({
    name: i18n.t('updatePlan:planItemNamePlaceholder'),
    studentId: plan.studentId,
    planId: plan.id,
    type,
    completed: false
  });