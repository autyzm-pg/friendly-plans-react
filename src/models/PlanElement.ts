import { PlanItemType } from './PlanItem';

export interface PlanElement {
  id: string;
  name: string;
  type: PlanItemType;
  completed: boolean;
  time: number;
  lector: boolean;

  complete: () => void;
  update: (changes: any) => void;
}
