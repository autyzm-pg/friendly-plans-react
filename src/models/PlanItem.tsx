import { RNFirebase } from 'react-native-firebase';
import { i18n } from '../locale';
import { getPlanItemRef, getPlanItemsRef, getPlanSubItemsRef } from './FirebaseRefProxy';
import { Plan } from './Plan';
import { PlanElement } from './PlanElement';
import { PlanSubItem } from './PlanSubItem';
import { ParameterlessConstructor, SubscribableModel } from './SubscribableModel';

export enum PlanItemType {
  SimpleTask = 'simpleTask',
  ComplexTask = 'complexTask',
  Break = 'break',
  Interaction = 'interaction',
  SubElement = 'subElement',
}

const PLAN_ITEMS_ICONS = {
  task: 'layers',
  break: 'bell',
  interaction: 'account-multiple',
};

export class PlanItem implements SubscribableModel, PlanElement {
  static create = (
    plan: Plan,
    type: PlanItemType,
    name: string = i18n.t('updatePlan:planItemNamePlaceholder'),
  ): Promise<RNFirebase.firestore.DocumentReference> =>
    getPlanItemsRef(plan.studentId, plan.id).add({
      name,
      studentId: plan.studentId,
      planId: plan.id,
      type,
      completed: false,
      lector: false,
      nameForChild: i18n.t('planItemActivity:taskNameForChild'),
    });

  static async createPlanItem(
    plan: Plan,
    type: PlanItemType,
    name: string = i18n.t('updatePlan:planItemNamePlaceholder'),
  ): Promise<PlanItem> {
    const { id } = await getPlanItemsRef(plan.studentId, plan.id).add({
      name,
      studentId: plan.studentId,
      planId: plan.id,
      type,
      completed: false,
      lector: false,
      nameForChild: i18n.t('planItemActivity:taskNameForChild'),
    });

    return Object.assign(new PlanItem(), {
      id,
      name,
      studentId: plan.studentId,
      planId: plan.id,
      type,
      completed: false,
      lector: false,
      nameForChild: i18n.t('planItemActivity:taskNameForChild'),
    });
  }

  id!: string;
  name!: string;
  type!: PlanItemType;
  planId!: string;
  studentId!: string;
  completed!: boolean;
  time!: number;
  image!: string;
  lector!: boolean;
  nameForChild!: string;

  getIconName = (): string => {
    return PLAN_ITEMS_ICONS[this.type];
  };

  isTask = (): boolean => this.type === PlanItemType.SimpleTask || this.type === PlanItemType.ComplexTask;
  isSimpleTask = (): boolean => this.type === PlanItemType.SimpleTask;
  complete = () => {
    this.update({ completed: true });
  };

  changeType = (type: PlanItemType) => {
    this.update({
      type,
    });
  };

  update = (changes: object) => getPlanItemRef(this.studentId, this.planId, this.id).update(changes);
  delete = (): Promise<void> => getPlanItemRef(this.studentId, this.planId, this.id).delete();

  getChildCollectionRef: () => RNFirebase.firestore.CollectionReference = () =>
    getPlanSubItemsRef(this.studentId, this.planId, this.id);
  getChildType: () => ParameterlessConstructor<SubscribableModel> = () => PlanSubItem;
  getRef: () => RNFirebase.firestore.DocumentReference = () => getPlanItemRef(this.studentId, this.planId, this.id);
}
