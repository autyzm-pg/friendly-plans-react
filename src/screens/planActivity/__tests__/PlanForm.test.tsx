import { shallow } from 'enzyme';
import React from 'react';

import { navigationMock } from 'helpers';
import { Plan } from 'models';
import { PlanForm } from '../PlanForm';

describe('PlanFormControls', () => {
  const props = {
    navigation: navigationMock,
    onSubmit: jest.fn(),
    plan: new Plan(),
  };

  it('renders correctly', () => {
    const planInput = shallow(<PlanForm {...props} />);

    expect(planInput).toMatchSnapshot();
  });
});
