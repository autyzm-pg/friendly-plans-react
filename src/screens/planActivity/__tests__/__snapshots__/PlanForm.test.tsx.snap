// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanFormControls renders correctly 1`] = `
<Formik
  enableReinitialize={false}
  initialValues={
    Object {
      "emoji": undefined,
      "planInput": undefined,
    }
  }
  isInitialValid={false}
  onSubmit={[MockFunction]}
  render={[Function]}
  validate={[MockFunction]}
  validateOnBlur={true}
  validateOnChange={true}
/>
`;
